source/common/stats/thread_local_store.cc

```diff
@@ -151,17 +148,21 @@ Counter& ThreadLocalStoreImpl::ScopeImpl::counter(const std::string& name) {
 
 void ThreadLocalStoreImpl::ScopeImpl::deliverHistogramToSinks(const std::string& name,
                                                               uint64_t value) {
-  std::string final_name = prefix_ + name;
-  for (Sink& sink : parent_.timer_sinks_) {
-    sink.onHistogramComplete(final_name, value);
+  if (!parent_.shutting_down_) {
+    std::string final_name = prefix_ + name;
+    for (Sink& sink : parent_.timer_sinks_) {
+      sink.onHistogramComplete(final_name, value);
+    }
   }
 }
 
 void ThreadLocalStoreImpl::ScopeImpl::deliverTimingToSinks(const std::string& name,
                                                            std::chrono::milliseconds ms) {
-  std::string final_name = prefix_ + name;
-  for (Sink& sink : parent_.timer_sinks_) {
-    sink.onTimespanComplete(final_name, ms);
+  if (!parent_.shutting_down_) {
+    std::string final_name = prefix_ + name;
+    for (Sink& sink : parent_.timer_sinks_) {
+      sink.onTimespanComplete(final_name, ms);
```

Comment:
Previously, we:
1) Did not delete any of the thread local objects. shutdownThread() just iterated through and called shutdown() on each object.
2) We would use shutdown() in the stat sink to not do any more flushing, which might make connections. Note however that this I think was already busted, where this would happen after cluster manager thread local shutdown(), but creating a tcp connection is stateless and safe.

Now, we actually do proper RAII cleanup, in a well defined (reverse) order. However, there is still a scenario in which during cluster manager thread local deletion, we may try to flush thread local timers/histograms. The problem here is that the stat flushing my try to create a new thread local TCP connection, creating circular breakage. No ordering will fix this. So I just special cased this case since I think that is simpler than trying to bring back some complex shutdown sequence (one other solution is to bring back shutdown() and do a first pass on that, then do deletion). If you think doing two passes with shutdown() would be better I can do that and then we don't need this guard, but then everyone has to implement shutdown() again which they won't need.

Anyway LMK what you think about ^. I'm not really sure there is much point in putting what came before in a comment either but I can at least put some comments on on why the guard is there now if we stick with this approach.